docs/design/upgrade.md

```diff
@@ -0,0 +1,203 @@
+Upgrade steps must have 3 properties:
+
+* **Atomicity** ensures that each step is "all or nothing"; if one part of the step fails, the entire step should fail, and the source is left unchanged.
+* **Consistency** ensures that each step will take the source from one compilable state to another compilable state. If the state is inconsistent before the step is applied (ie. it does not compile), the step should not be applied and/or fail.
+* **Durability** ensures that once a step succeeds, it is permanent and should be committed, even in the event of a power loss. This implies that the final operation should be to commit the code to a VCS.
+
+One might figure that, adding an Isolation property, this would match the ACID principle of database transactions. Since multiple steps cannot be performed in parallel, the source is already isolated and it would be a redundant property.
+
+Please note that because of the consistency of a step, assuming the property that code compile in one version, it is imperative that code compiles in the version after the step is applied. This leads to a strong induction with regards to the "can compile" property.
+
+If we add another property; invertibility, we could technically _undo_ upgrade steps and as such, if an error is inserted in the process, we could revert the step, make the corrections, then apply the step again.
+
+### **RFC: should invertibility be a requirement or ignored.**
+
+## Step Definition
+
+In the CLI, the `upgrade/steps/` directory will be kept for step scripts. It has the following structure (example):
+
+    angular-cli/
+      upgrade/
+        index.ts                    <-- the Upgrade api (see below).
+        steps/
+          angular2/                 <-- or other npm packages.
+            0001/                   <-- incrementing number.
+              index.ts              <-- the script which will be executed to move up.
+              XYZ.ts                <-- some utility used by a step.
+
+The `index.ts` of the step exports the following symbols:
+
+* `upgrade(project: AngularProject): Observable<Change>`; upgrade from the current version to the `toVersion`. Returns a stream of changes.
+* `downgrade(project: AngularProject): Observable<Change>`; downgrade from the `toVersion` to the version of this script. Returns a stream of changes.
+* `fromVersion: SemVersion | SemVersion[]`; either a version string or an array of all version strings representing the versions this package must be before `up()` is called.
+* `toVersion: SemVersion`; a version string representing the current version when `up()` succeeds.
+
+### Rationales
+An incrementing number is used instead of the from version because some versions might not be a 1-to-1 relationship between two versions. For example, version `beta-12` and `beta-13` might be both no-op, and the script upgrading to `beta-14` might have a `fromVersion` of `[beta-12, beta-13]`. Using a binary search, the lookup time is not a factor.
```

Comment:
This is a tough one, because the semver does not give us an easy way to know that `1.0.0.alpha.4` > `1.0.0.beta.2`. So we need to make the mapping an `N:1` relationship, unfortunately.